### PR TITLE
feat: Add wasm_memory_threshold to ProposeToUpdateCanisterSettingsCmd

### PR DESCRIPTION
**Pull request by dragoljub-duric**
To support `on_low_wasm_memory` hook we are adding `wasm_memory_threshold` to `NNS` so it allows users to update its value using `NNS` proposal. We are as well adding `wasm_memory_threshold` to canister settings to canisters in `IC` repo.

The `wasm_memory_threshold` will be added to the interface specification in https://github.com/dfinity/portal/pull/3761

### CODE REVIEW

File: rs/nns/governance/src/proposals/update_canister_settings.rs [hunk 1]
```diff
@@ -75,6 +76,7 @@ impl UpdateCanisterSettings {
         let memory_allocation = settings.memory_allocation.map(Nat::from);
         let freezing_threshold = settings.freezing_threshold.map(Nat::from);
         let wasm_memory_limit = settings.wasm_memory_limit.map(Nat::from);
+        let _wasm_memory_threshold = settings.wasm_memory_threshold.map(Nat::from);
```
Comment (dragoljub-duric) [inline rs/nns/governance/src/proposals/update_canister_settings.rs:79]: Should we add this field to RootCanisterSettings?

Comment (aterga) [reply rs/nns/governance/src/proposals/update_canister_settings.rs:79]: Yes, but it could be done in a follow-up PR ideally

File: rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs [hunk 2]
```diff
@@ -2570,6 +2570,7 @@ pub mod update_canister_settings {
         pub log_visibility: Option<i32>,
         #[prost(uint64, optional, tag = "6")]
         pub wasm_memory_limit: Option<u64>,
+        /// XXX HOW TO RUN PROTOBUF GENERATOR TO UPDATE THIS STRUCT?
```
[the remark this answers is not in the document]
Comment (aterga) [reply rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs:2573]: This file is not generated via proto, you should edit it manually.

In the future, 
rs/nns/governance/src/gen/ic_nns_governance.pb.v1.rs
and 
rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs

may diverge.

**Review by aterga — APPROVED** [review]

File: rs/nns/governance/src/proposals/update_canister_settings.rs (hunk 1 above)
Comment (dragoljub-duric) [reply rs/nns/governance/src/proposals/update_canister_settings.rs]: Thank you. I added it manually. I also assumed that the generator would add fields only inside `/gen/` files, while I manually added them to the rest.

File: rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dragoljub-duric) [reply rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs]: Thank you I added it manually. I also supposed that only files in `/gen/` directories are generated, and in the non generated, I added copies of changed lines form respective generated files.

File: packages/pocket-ic/src/management_canister.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ pub struct DefiniteCanisterSettings {
     pub wasm_memory_limit: candid::Nat,
     pub memory_allocation: candid::Nat,
     pub compute_allocation: candid::Nat,
+    pub wasm_memory_threshold: candid::Nat,
```
Comment (mraszyk) [inline packages/pocket-ic/src/management_canister.rs:31]: this file should stay in sync with the interface spec so this shouldn't be added before publicly releasing the feature

Comment (dragoljub-duric) [reply packages/pocket-ic/src/management_canister.rs]: Thanks, done.

File: rs/bitcoin/ckbtc/minter/ckbtc_minter.did [hunk 4]
```diff
@@ -29,6 +29,7 @@ type DefiniteCanisterSettings = record {
   reserved_cycles_limit : nat;
   log_visibility: LogVisibility;
   wasm_memory_limit : nat;
+  wasm_memory_threshold : nat;
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/ckbtc_minter.did:32]: Here for a lack of a better place: could you add a PR description to motivate this change?

**Review by gregorydemay — APPROVED** [review]

File: rs/bitcoin/ckbtc/minter/ckbtc_minter.did (hunk 4 above)
Comment (dragoljub-duric) [reply rs/bitcoin/ckbtc/minter/ckbtc_minter.did:32]: @gregorydemay It seems Bitcoin and Ethereum tests are failing because of the incompatible candid interface, but I do not know what should be changed. Do you have any idea how to fix it?

**Comment by dragoljub-duric** [comment]
@dfinity/nns-team Can someone please help me, I cannot find why test `//rs/nervous_system/integration_tests:integration_tests_test_tests/sns_release_qualification` is failing?

File: rs/nns/integration_tests/src/update_canister_settings.rs [hunk 5]
```diff
@@ -36,7 +36,8 @@ fn test_update_canister_settings_proposal(
     let target_memory_allocation = 1u64 << 33;
     let target_compute_allocation = 10u64;
     let target_freezing_threshold = 100_000u64;
-    let target_wasm_memory_limit = 1u64 << 32;
+    let target_wasm_memory_limit = 1u64 << 36;
```
Comment (aterga) [inline rs/nns/integration_tests/src/update_canister_settings.rs:39]: Why was this changed? 

```suggestion
    let target_wasm_memory_limit = 1u64 << 32;
```

Comment (dragoljub-duric) [reply rs/nns/integration_tests/src/update_canister_settings.rs:39]: target_wasm_memory_limit needs to be smaller than `target_wasm_memory_threshold`. I also wanted to test whether `target_wasm_memory_threshold` will work with values greater than `u32`, so I needed this.

File: rs/bitcoin/ckbtc/minter/ckbtc_minter.did (hunk 4 above)
Comment (gregorydemay) [reply rs/bitcoin/ckbtc/minter/ckbtc_minter.did]: > I just reverted changes on your side, in the case you want to use the feature, you can add the new field, for now, I believe it is unnecessary.

It would be nice to have it is this is exposed. Do you have the exact error that occurred? We are working on that area, so it could be that the error is due to an outdated master (merging master should fix it)

File: rs/nns/integration_tests/src/update_canister_settings.rs (hunk 5 above)
Comment (aterga) [reply rs/nns/integration_tests/src/update_canister_settings.rs:39]: This justification should IMHO be written in a comment.

Comment (dragoljub-duric) [reply rs/nns/integration_tests/src/update_canister_settings.rs:40]: Done.

**Comment by anchpop** [comment]
> @dfinity/nns-team Can someone please help me, I cannot find why test `//rs/nervous_system/integration_tests:integration_tests_test_tests/sns_release_qualification` is failing?

I've pushed two commits, and it should pass now

File: rs/sns/swap/canister/swap.did [hunk 6]
```diff
@@ -53,6 +53,7 @@ type DefiniteCanisterSettingsArgs = record {
   wasm_memory_limit : opt nat;
   memory_allocation : nat;
   compute_allocation : nat;
+  wasm_memory_threshold : nat;
```
Comment (aterga) [inline rs/sns/swap/canister/swap.did:56]:
```suggestion
  wasm_memory_threshold : opt nat;
```
